cmd/private-org-peribolos-sync/main.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"k8s.io/test-infra/prow/config/org"
+	"k8s.io/test-infra/prow/config/secret"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"github.com/openshift/ci-tools/pkg/promotion"
+)
+
+type gitHubClient interface {
+	GetRepo(owner, name string) (github.FullRepo, error)
+}
+
+type options struct {
+	peribolosConfig string
+	destOrg         string
+	releaseRepoPath string
+	whitelist       flagutil.Strings
+	github          flagutil.GitHubOptions
+
+	whitelistByOrg map[string]sets.String
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.peribolosConfig, "peribolos-config", "", "Peribolos configuration file")
+	fs.StringVar(&o.releaseRepoPath, "release-repo-path", "", "Path to a openshift/release repository directory")
+	fs.StringVar(&o.destOrg, "destination-org", "", "Destination name of the peribolos configuration organzation")
+
+	fs.Var(&o.whitelist, "whitelist", "One or more repositories that do not promote official images and need to be included")
+
+	o.github.AddFlags(fs)
+	fs.Parse(os.Args[1:])
+
+	return o
+}
+
+func validateOptions(o *options) error {
+	var validationErrors []error
+
+	if len(o.releaseRepoPath) == 0 {
+		validationErrors = append(validationErrors, errors.New("--release-repo-path is not specified"))
+	}
+	if len(o.peribolosConfig) == 0 {
+		validationErrors = append(validationErrors, errors.New("--peribolos-config is not specified"))
+	}
+	if len(o.destOrg) == 0 {
+		validationErrors = append(validationErrors, errors.New("--destination-org is not specified"))
+	}
+
+	if len(o.whitelist.Strings()) > 0 {
+		if err := validateWhitelist(o.whitelist.Strings()); err != nil {
+			validationErrors = append(validationErrors, err)
+		} else {
+			o.whitelistByOrg = getWhitelistByOrg(o.whitelist.Strings())
+		}
+	}
+
+	if err := o.github.Validate(false); err != nil {
+		validationErrors = append(validationErrors, err)
+	}
+
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validateWhitelist(whitelist []string) error {
+	for _, repo := range whitelist {
+		if len(strings.Split(repo, "/")) != 2 {
+			return fmt.Errorf("--whitelist: %s must be in org/repo format", repo)
+		}
+	}
+	return nil
+}
+
+func getWhitelistByOrg(whitelist []string) map[string]sets.String {
+	ret := make(map[string]sets.String)
+
+	for _, repo := range whitelist {
+		orgRepo := strings.Split(repo, "/")
+
+		orgName := orgRepo[0]
+		repoName := orgRepo[1]
+
+		ret[orgName] = ret[orgName].Union(sets.NewString(repoName))
```

Comment:
err... `Insert(repoName)`?

Reply:
It will panic. because there is a chance that the `ret[orgName]` won't exist.

Reply:
Why does `ret[orgName].Union()` work but `ret[orgName].Insert()` not? This is silly. Just do

```go
repos, exist := ret[orgName]
if !exist {
    repos = sets.NewString()
}
repos.Insert(repoName)
```

The code as written stands out as not normal and makes you wonder why we're doing what we're doing. We want to do a set insertion, so we should do that. By handling the `nil` case explicitly, in code, we make our intent clear to the reader.

Reply:
I had this like this in the previous version of the commit, but it seems `Union` does all that in one line.  However, I am willing to rewrite it as you suggested because is more readable.

Reply:
addressed

Reply:
The reasoning here is that code is built to be read, not written. There is no utility to us if it is one line or not. Using a side-effect of `Union()` implementation to allow us to do it in one line makes it (maybe) easier to write since there's less code but _definitely_ makes it harder to read. We always want to prefer to make it readable.